src/Illuminate/Container/BoundMethod.php

```diff
@@ -127,42 +134,47 @@ protected static function getMethodDependencies($container, $callback, array $pa
      *
      * @throws \ReflectionException
      */
-    protected static function getCallReflector($callback)
+    protected static function getCallReflector($callback): \Reflector
     {
         if (is_string($callback) && strpos($callback, '::') !== false) {
             $callback = explode('::', $callback);
         }
 
         return is_array($callback)
-                        ? new ReflectionMethod($callback[0], $callback[1])
-                        : new ReflectionFunction($callback);
+            ? new ReflectionMethod($callback[0], $callback[1])
+            : new ReflectionFunction($callback);
     }
 
     /**
-     * Get the dependency for the given call parameter.
+     * Add the dependencies to the input data.
      *
-     * @param  \Illuminate\Container\Container  $container
-     * @param  \ReflectionParameter  $parameter
-     * @param  array  $parameters
-     * @param  array  $dependencies
-     * @return void
+     * @param  \Illuminate\Container\Container $container
+     * @param  array $signature
+     * @param  array $inputData
+     * @return array
      */
-    protected static function addDependencyForCallParameter($container, $parameter,
-                                                            array &$parameters, &$dependencies)
+    protected static function addDependencyForCallParameter($container, array $signature, array $inputData): array
     {
-        if (array_key_exists($parameter->name, $parameters)) {
-            $dependencies[] = $parameters[$parameter->name];
-
-            unset($parameters[$parameter->name]);
-        } elseif ($parameter->getClass() && array_key_exists($parameter->getClass()->name, $parameters)) {
-            $dependencies[] = $parameters[$parameter->getClass()->name];
-
-            unset($parameters[$parameter->getClass()->name]);
-        } elseif ($parameter->getClass()) {
-            $dependencies[] = $container->make($parameter->getClass()->name);
-        } elseif ($parameter->isDefaultValueAvailable()) {
-            $dependencies[] = $parameter->getDefaultValue();
+        $resolvedInputData = [];
+        $i = 0;
+
+        foreach ($signature as $parameter) {
+            if (array_key_exists($parameter->name, $inputData)) {
+                $resolvedInputData[] = $inputData[$parameter->name];
+            } elseif ($parameter->getClass() && array_key_exists($parameter->getClass()->name, $inputData)) {
+                $resolvedInputData[] = $inputData[$parameter->getClass()->name];
+            } elseif ($parameter->getClass()) {
+                $resolvedInputData[] = $container->make($parameter->getClass()->name);
+            } elseif (isset($inputData[$i])) {
+                $data = $inputData[$i];
+                $i++;
+                $resolvedInputData[] = $data;
```

Comment:
I think you can just do `$resolvedInputData[] = $inputData[$i]` here and not have `$data`?

Reply:
That part will change, although all tests are passing by now, I am pretty well sure that there is a bug sneaking us.